### DISCUSSION TITLE
How to execute a Task when a new Item is added to a queue

### DISCUSSION THREAD

**Discussion by obivandamme** [OrchardCMS/OrchardCore#11971] [Q&A]
Hi all,

**Background**
I am working on an application that synchonizes multiple systems and so far we used workflow activities on a workflow that is triggered every two hours. Sometimes it happens that one of the activities needs to be triggered by hand with a diffferent set of parameters than the configured workflow activity. This is done by having additional workflows with an http trigger that only consist of the event and a single activity. 

Unfortunately these synchonization tasks are not allowed to run in parallel. So every time someone wants to run a job by hand he/she needs to make sure the timed workflow is not running. 

We tried to solve this issue by having a queue where each time a new item is added, a synchronization task is started and than have the manual triggers and the timer simply add new items to the queue. The important part now is that the queue consumer processes the items one after the other to prevent any tasks from running in parallel. The Idea was to have a background task that gets all queue items and executes the sync tasks. 

**Question**
How can I make sure that the DoWorkAsync method of IBackground task is not called again as long as the first call is not finished?

**Comment by jtkech** [comment]
For an `IBackgroundTask` you can define 2 locking parameters, by code with the `[BackgroundTask()]` attribute or through the admin dashboard if you enable the `OC.BackgroundTasks` feature. You can tweak these 2 parameter values for your need, there is the `LockTimeout`, the max time to acquire a lock, and the `LockExpiration`. If multiple instances of your application are running, you may need a distributed lock as we provide oob with the `Redis.Lock` feature.

A Workflow type also has the same locking parameters to ensure that 2 instances of the same WF type are not executed concurrently, but when an instance stop to execute the WF on a blocking activity, another instance can execute the WF, just to say that it doesn't ensure the atomicity of the whole WF. If you need it, in addition to the locking values there is another property `Singleton` to ensure that no other instance starts to execute the WF before it is fully completed.

Hmm but yes, to only use the solution at the WF level your manual and timed WFs should trigger the same WF type on which you would define locking values (both > 0), not sure that a WF can trigger directly another WF, at least it is possible through the http request/event/signal activities.

**Comment by obivandamme** [comment]
For the `IBackgroundTask` I only need to configure the Lock if more than one Instance of my application are running, right? If not, registering the Task as singleton on startup should do the trick. 
But let's say I want to future proof and already set these two lock parameters. What values should these have in case I don't know the exact execution times of a single call to `DoWorkAsync`, or the execution times vary? 
Will the lock get released when execution is done, even though the `LockExpiration` was not reached? If so, I expect that I have to say `LockTimeout` to a value shorter than trigger interval and `LockExpiration` to a value greater than the expected maximum run time of a single call to `DoWorkAsync`. Is that correct?

**Comment by jtkech** [comment]
> For the IBackgroundTask I only need to configure the Lock if more than one Instance of my application are running, right?

Yes

> If not, registering the Task as singleton on startup should do the trick.

Sorry, when I talked about Singleton I was not talking about how you register your IBackgroundTask  in the DI, I was talking about WF type property that you can set through the admin UI.

> Will the lock get released when execution is done, even though the LockExpiration was not reached?

Yes

>  If so, I expect that I have to say LockTimeout to a value shorter than trigger interval

Normally << 1 min the granularity of our BackGroundService. Then it depends on your need, the higher it is, the more the waiting task has a chance to be executed. The lower it is, the more the executing task becomes exclusive.

If you really want your tasks to be executed by only one running instance, you may be interested by the comments of this issue #11463.

> and LockExpiration to a value greater than the expected maximum run time of a single call to DoWorkAsync. Is that correct?

Yes, but normally still << 1 min the granularity of our BackGroundService.
